hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/test/java/org/apache/hadoop/yarn/client/api/impl/TestNMClient.java

```diff
@@ -125,576 +117,362 @@ public void postTransition(
-      testCommitContainer(container.getId(), false);
+      testReInitializeContainer(container, clc, false);
+      testContainerStatusRunning(container,  "will be Re-initialized");
+      waitForContainerRunningTransitionCount(container, 3);
+      testContainerRollback(container, true);
+      testContainerStatusRunning(container, "will be Rolled-back");
+      waitForContainerRunningTransitionCount(container, 4);
+      testContainerCommit(container, false);
+      testReInitializeContainer(container, clc, false);
+      testContainerStatusRunning(container, "will be Re-initialized");
+      waitForContainerRunningTransitionCount(container, 5);
+      testContainerCommit(container, true);
     } else {
-      testReInitializeContainer(container.getId(), clc, true);
-      testGetContainerStatus(container, i, ContainerState.RUNNING,
-          "will be Re-initialized", exitCode);
-      waitForContainerTransitionCount(container,
-          org.apache.hadoop.yarn.server.nodemanager.
-              containermanager.container.ContainerState.RUNNING, 3);
-      testRollbackContainer(container.getId(), true);
-      testCommitContainer(container.getId(), true);
+      testReInitializeContainer(container, clc, true);
+      testContainerStatusRunning(container, "will be Re-initialized");
+      waitForContainerRunningTransitionCount(container, 3);
+      testContainerRollback(container, false);
+      testContainerCommit(container, false);
     }
+    client.stopContainer(container.getId(), container.getNodeId());
+    testContainerStatusCompleted(container, "killed by the ApplicationMaster");
+  }
 
-    try {
-      client.stopContainer(container.getId(), container.getNodeId());
-    } catch (YarnException e) {
-      throw (AssertionError)
-        (new AssertionError("Exception is not expected: " + e, e));
+  private void waitForContainerRunningTransitionCount(Container container, long transitions) {
+    while (DebugSumContainerStateListener.RUNNING_TRANSITIONS
+        .getOrDefault(container.getId(), 0) != transitions) {
+      sleep(500);
     }
+  }
 
-    // getContainerStatus can be called after stopContainer
-    try {
-      // O is possible if CLEANUP_CONTAINER is executed too late
-      // -105 is possible if the container is not terminated but killed
-      testGetContainerStatus(container, i, ContainerState.COMPLETE,
-          "Container killed by the ApplicationMaster.",
-          Arrays.asList(
-              ContainerExitStatus.KILLED_BY_APPMASTER,
-              ContainerExitStatus.SUCCESS));
-    } catch (YarnException e) {
-      // The exception is possible because, after the container is stopped,
-      // it may be removed from NM's context.
-      if (!e.getMessage()
-            .contains("was recently stopped on node manager")) {
-        throw (AssertionError)
-          (new AssertionError("Exception is not expected: ", e));
-      }
+  private void testContainerStatusRunning(
```

Comment:
TBH, I liked the previous generic version (called testGetContainerStatus) better.
Now, testContainerStatusRunning and testContainerStatusCompleted have all the logic duplicated.
Not to mention that for every new state, you need yet another method to copy.
I suggest to make a generic version of this "container state checking" method or bring back the original.

Reply:
For 2 state i dont think we need generic logic, based on KISS, now you can read all the hacks in 1-1 methods